ext/network.go

```diff
@@ -0,0 +1,481 @@
+//	isCIDR(string) -> bool
+//
+// Examples:
+//
+//	cidr('192.168.0.0/24')
+//	cidr('::1/128')
+//	isCIDR('10.0.0.0/8') // true
+//
+// # IP Inspection and Canonicalization
+//
+// IP objects support various inspection methods.
+//
+//	<ip>.family() -> int
+//	<ip>.isLoopback() -> bool
+//	<ip>.isGlobalUnicast() -> bool
+//	<ip>.isLinkLocalMulticast() -> bool
+//	<ip>.isLinkLocalUnicast() -> bool
+//	<ip>.isUnspecified() -> bool
+//
+// The `ip.isCanonical` function takes a string and returns true if it matches
+// the RFC 5952 canonical string representation of that address.
+//
+//	ip.isCanonical(string) -> bool
+//
+// Examples:
+//
+//	ip('127.0.0.1').family() == 4
+//	ip('::1').family() == 6
+//	ip('127.0.0.1').isLoopback() == true
+//	ip.isCanonical('2001:db8::1') == true  // RFC 5952 format
+//	ip.isCanonical('2001:DB8::1') == false // Uppercase is not canonical
+//	ip.isCanonical('2001:db8:0:0:0:0:0:1') == false // Expanded is not canonical
+//
+// # CIDR Member Functions
+//
+// CIDR objects support containment checks and property extraction.
+//
+//	<cidr>.containsIP(ip|string) -> bool
+//	<cidr>.containsCIDR(cidr|string) -> bool
+//	<cidr>.ip() -> ip
+//	<cidr>.masked() -> cidr
+//	<cidr>.prefixLength() -> int
+//
+// Examples:
+//
+//	cidr('10.0.0.0/8').containsIP(ip('10.0.0.1')) == true
+//	cidr('10.0.0.0/8').containsIP('10.0.0.1') == true
+//	cidr('10.0.0.0/8').containsCIDR('10.1.0.0/16') == true
+//	cidr('192.168.1.5/24').ip() == ip('192.168.1.5')
+//	cidr('192.168.1.5/24').masked() == cidr('192.168.1.0/24')
+//	cidr('192.168.1.0/24').prefixLength() == 24
+func Network() cel.EnvOption {
+	return cel.Lib(&networkLib{})
+}
+
+const (
+	// Function names
+	isIPFunc             = "isIP"
+	isCIDRFunc           = "isCIDR"
+	ipFunc               = "ip"
+	cidrFunc             = "cidr"
+	familyFunc           = "family"
+	isCanonicalFunc      = "ip.isCanonical"
+	isLoopbackFunc       = "isLoopback"
+	isGlobalUnicastFunc  = "isGlobalUnicast"
+	isUnspecifiedFunc    = "isUnspecified"
+	isLinkLocalMcastFunc = "isLinkLocalMulticast"
+	isLinkLocalUcastFunc = "isLinkLocalUnicast"
+	containsIPFunc       = "containsIP"
+	containsCIDRFunc     = "containsCIDR"
+	maskedFunc           = "masked"
+	prefixLengthFunc     = "prefixLength"
+	ipFromCIDRFunc       = "ip"
+)
+
+var (
+	// Definitions for the Opaque Types
+	networkIPType   = types.NewTypeValue("network.IP", traits.ReceiverType)
+	networkCIDRType = types.NewTypeValue("network.CIDR", traits.ReceiverType)
+)
+
+type networkLib struct{}
+
+func (*networkLib) LibraryName() string {
+	return "cel.lib.ext.network"
+}
+
+func (*networkLib) CompileOptions() []cel.EnvOption {
+	return []cel.EnvOption{
+		// 1. Register Types
+		cel.Types(
+			networkIPType,
+			networkCIDRType,
+		),
+		// 2. Register Adapter (Bundled here so it applies automatically)
+		cel.CustomTypeAdapter(&networkAdapter{
+			Adapter: types.DefaultTypeAdapter,
+		}),
+		// 3. Register Functions
+		cel.Function(isIPFunc,
```

Comment:
I would be very happy if we used the K8s content verbatim here (modulo some simple renames) as it'll simplify cost-tracking checks which K8s is doing presently and which someone will also need to upstream before making the improvements available to K8s to use.

https://github.com/kubernetes/kubernetes/blob/5bcb7599736327cd8c6d23e398002354a6e40f68/staging/src/k8s.io/apiserver/pkg/cel/library/ip.go

The primary cleanup would be to shift from an `IP` library to a `Network` library which aggregates the K8s `library/ip.go` and `library/cidr.go` since we'd very likely always use them together.